contracts-bedrock: move immutables to storage for L1CrossDomainMessenger

0xfuturistic opened the pull request:
context: https://github.com/ethereum-optimism/client-pod/issues/265
security review (wip): https://www.notion.so/oplabs/DRAFT-Security-Review-Move-Immutables-to-Storage-in-L1-Contracts-56c07789fa454d8fb2df80f0b84a483b?pvs=4
legacy PR: https://github.com/ethereum-optimism/optimism/pull/6574

0xfuturistic committed:
contracts-bedrock: move immutables to storage in L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update Deploy script to support changes to L1CrossDomainMessenger

L1CrossDomainMessenger had its immutables moved to storage

0xfuturistic committed:
contracts-bedrock: update tests to support changes to L1CrossDomainMessenger

L1CrossDomainMessenger had its immutables moved to storage

0xfuturistic committed:
contracts-bedrock: fix Deploy script for L1CrossDomainMessenger

Since values are set at initialization and not at deployment, ChainAssertions.checkL1CrossDomainMessenger reverts if called at this point with the dependent contracts. Instead, this call should be done immediately after initialization (as is already done in the Deploy script).

0xfuturistic committed:
op-chain-ops: update upgrade script for L1CrossDomainMessenger

L1CrossDomainMessenger had its immutable variables moved to storage

0xfuturistic committed:
contracts-bedrock: fix L1CrossDomainMessenger spec in Authorization.t.sol

Addresses the fact that L1CrossDomainMessenger's spec for initialize should include the parameters for the variables moved from immutables into storage

0xfuturistic committed:
contracts-bedrock: update semver for L1CrossDomainMessenger

Inphi commented:
You should run `pnpm pre-pr` in contracts/bedrock to fix the `contract-bedrock-checks` build.

Inphi commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -467,20 +467,20 @@ contract Deploy is Deployer {
-            new L1CrossDomainMessenger{ salt: _implSalt() }({ _portal: OptimismPortal(payable(portal)) });
+        L1CrossDomainMessenger messenger = new L1CrossDomainMessenger{ salt: _implSalt() }();
+
+        // TODO: address inconsistency in variable names vs. type in PORTAL() vs. portal() in L1CrossDomainMessenger
```
this is for backwards-compatibility iirc. We used to use uppercase for public immutables, and now they're hidden behind explicit getter functions.

Inphi commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -467,20 +467,20 @@ contract Deploy is Deployer {
-        // are always proxies.
-        Types.ContractSet memory contracts = _proxiesUnstrict();
-        contracts.L1CrossDomainMessenger = address(messenger);
-        ChainAssertions.checkL1CrossDomainMessenger({ _contracts: contracts, _vm: vm, _isProxy: false });
```
I don't think this should be removed. `ChainAssertions` are a modular set of post-deployment checks that are used by multiple deploy scripts. So it's better to make changes there, rather than manually do them in a deploy script, so that all deploy scripts use the same checks.

Instead, you should modify `ChainAssertions.checkL1CrossDomainMessenger` to check that the `portal` was initialized correctly (for both the proxy and actual contract).
In here, where we're deploying the contract itself, the portal should be `0`. For the proxy, post-initialization, it should be the `optimismPortalProxy`.

0xfuturistic replied:
good call! I like this idea. will implement

Inphi commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -467,20 +467,20 @@ contract Deploy is Deployer {
+        // TODO: check that the lines below make sense
+        //       (gotten from https://github.com/ethereum-optimism/optimism/pull/6574)
+        bytes32 xdmSenderSlot = vm.load(address(messenger), bytes32(uint256(204)));
+        require(address(uint160(uint256(xdmSenderSlot))) == Constants.DEFAULT_L2_SENDER);
```
This seems redundant (and incorrect). The xdmSender isn't set up at deployment, only during initialization. So the expected slot value should be 0 here, not `Constants.DEFAULT_L2_SENDER`. Besides, the existing `ChainAssertions.checkL1CrossDomainMessenger` call already checks the xdmSender slot.

Inphi commented on packages/contracts-bedrock/scripts/Deploy.s.sol:
```diff
@@ -885,6 +887,14 @@ contract Deploy is Deployer {
+        // TODO: check that the lines below make sense
+        //       (gotten from https://github.com/ethereum-optimism/optimism/pull/6574)
+        bytes32 xdmSenderSlot = vm.load(address(messenger), bytes32(uint256(204)));
+        require(address(uint160(uint256(xdmSenderSlot))) == Constants.DEFAULT_L2_SENDER);
```
redundant. This should be handled by `ChainAssertions.checkL1CrossDomainMessenger` on L885. And `ChainAssertions.checkL1CrossDomainMessenger` should be tweaked to check the l1 xdm `portal`.

Inphi commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -13,50 +13,53 @@ import { SuperchainConfig } from "src/L1/SuperchainConfig.sol";
-    OptimismPortal public immutable PORTAL;
+    /// @notice Address of the OptimismPortal.
+    /// @custom:network-specific
+    OptimismPortal public portal;
```
This breaks ABI because it reuses the existing `superchainConfig` slot. Though (and maybe @tynes can correct me), since we have yet to deploy 2.1.x yet, this breaking change should be OK.

0xfuturistic replied:
ended up moving the variables after `superchainConfig` to keep the same slot for `superchainConfig`. confirmed by looking at the storage layout

0xfuturistic committed:
contracts-bedrock: move immutable variable in CrossDomainMessenger into storage

0xfuturistic committed:
contracts-bedrock: update L1CrossDomainMessenger to handle change to CrossDomainMessenger

CrossDomainMessenger had its immutable variable moved into storage

0xfuturistic committed:
contracts-bedrock: update L2CrossDomainMessenger to handle change to CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update Initializable.t.sol for changes made to l1CrossDomainMessenger, l2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update Deploy script to handle changes to L1CrossDomainMessenger

0xfuturistic committed:
op-chain-ops: redefine L2CrossDomainMessenger immutable config

immutable variables were removed

0xfuturistic committed:
op-chain-ops: introduce storage values for L2CrossDomainMessenger

corresponding to the immutable variable moved into storage in the contract

0xfuturistic committed:
contracts-bedrock: move otherMessenger var to after existing vars in L1CrossDomainMessenger

This helps maintain the same storage slot for the existing variables, which are used by other scripts through vm.load

0xfuturistic committed:
contracts-bedrock: fix test spec for Authorization for L1CrossDomainMessenger

was missing newly introduced parameter for variable moved into storage

0xfuturistic committed:
contracts-bedrock: reorder variables in L1CrossDomainMessenger

this is to maintain the storage slot of the existing variables (that is, the ones that weren't immutables before)

0xfuturistic committed:
contracts-bedrock: add parameter documentation to __CrossDomainMessenger_init

0xfuturistic committed:
contracts-bedrock: reintroduce ChainAssertions at deployment of L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: relabel and reorganize variables in upgrade script for L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: make snapshots for L1CrossDomainManager, L2CrossDomainManager

0xfuturistic committed:
contracts-bedrock: make slither report for contracts-bedrock: make snapshots for L1CrossDomainManager, L2CrossDomainManager

0xfuturistic committed:
contracts-bedrock: update semver-lock after changes to L1CrossDomainMessenger, L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: make gas snapshot

0xfuturistic committed:
op-chain-ops: fix missing parameter for check of L1CrossDomainMessenger initialization

0xfuturistic committed:
op-bindings: make for L1CrossDomainMessenger, L2CrossDomainMessenger, CrossDomainMessenger

tynes commented on packages/contracts-bedrock/snapshots/storageLayout/L1CrossDomainMessenger.json:
```diff
@@ -111,18 +111,32 @@
     "label": "superchainConfig",
     "offset": 0,
-    "slot": "251",
+    "slot": "252",
```
We cannot change this storage slot

tynes commented on packages/contracts-bedrock/snapshots/storageLayout/L2CrossDomainMessenger.json:
```diff
@@ -111,11 +111,18 @@
     "label": "__gap",
     "offset": 0,
-    "slot": "207",
+    "slot": "208",
```
We should make this type smaller

tynes commented on packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol:
```diff
@@ -114,9 +114,6 @@ abstract contract CrossDomainMessenger is
     ///         call in `relayMessage`.
     uint64 public constant RELAY_GAS_CHECK_BUFFER = 5_000;
```
Good thinking but no need to do that now

tynes commented on packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol:
```diff
@@ -183,7 +179,7 @@ abstract contract CrossDomainMessenger is
         // guarantee the property that the call to the target contract will always have at least
         // the minimum gas limit specified by the user.
         _sendMessage(
```
It would be amazing if you could update the calls in this contract to use the kwargs style syntax where its

```solidity
_sendMessage({
  key: value,
  ...
})
```

tynes replied:
This will help to reduce cognitive overhead during review. Prefer this style anytime there are 3 or more arguments or its not obvious what the arguments are

0xfuturistic committed:
contracts-bedrock: make snapshots for L1CrossDomainMessenger, L2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update slither report

0xfuturistic committed:
contracts-bedrock: update snapshots for L1CrossDomainMessenger, L2CrossDomainMessenger

0xfuturistic committed:
op-bindings: make for CrossDomainMessenger, L1CrossDomainMessenger, L2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: use isProxy instead of isInitialized for ChainAssertions of L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: fix spec

0xfuturistic committed:
contracts-bedrock: drop unused dependency in ChainAssertions

0xfuturistic committed:
contracts-bedrock: move variable otherMessenger to end of storage in CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: improve parameter labeling in L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: improve parameter labeling in L2CrossDomainMessenger

0xfuturistic committed:
op-chain-ops: use predeploys to access L2CrossDomainMessenger in L1CrossDomainMessenger upgrade script

0xfuturistic committed:
contracts-bedrock: update slither report

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
contracts-bedrock: update snapshots

0xfuturistic committed:
op-bindings: make for L1CrossDomainMessenger, L2CrossDomainMessenger, CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: add tests for L1CrossDomainMessenger

Tests for constructor and initializer values

0xfuturistic committed:
op-chain-ops: make otherMessenger dynamic in L1CrossDomainMessenger upgrade script

0xfuturistic committed:
contracts-bedrock: shorten storage gap in L1CrossDomainMessenger and move otherMessenger before it

this helps preserve the existing storage layout while moving otherMessenger from immutable into storage

0xfuturistic committed:
contracts-bedrock: make snapshots for L1CrossDomainMessenger

0xfuturistic committed:
op-bindings: make for CrossDomainMessenger, L1CrossDomainMessenger, L2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: improve documentation for CrossDomainMessenger, L1CrossDomainMessenger, L2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
contracts-bedrock: fix comments in CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: change type of legacy getter for L1CrossDomainMessenger PORTAL

0xfuturistic committed:
contracts-bedrock: add tests for L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: add tests for L2CrossDomainMessenger

0xfuturistic committed:
op-bindings: make for L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update snapshots

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
contracts-bedrock: reorder isProxy in ChainAssertions for L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: use predeploys in Deploy script for L1CrossDomainMessenger

0xfuturistic committed:
op-bindings: make for L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
contracts-bedrock: update snapshots

tynes commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -107,7 +108,26 @@ func L1CrossDomainMessenger(batch *safe.Batch, implementations superchain.Implem
-	calldata, err := l1CrossDomainMessengerABI.Pack("initialize", superchainConfigProxy)
+	var optimismPortal, otherMessenger common.Address
+	if config != nil {
+		optimismPortal = common.HexToAddress(list.OptimismPortalProxy.String())
```
Same comment as other PRs, prefer to pull remote values and then check them for equality against the config if the config is present

tynes commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -13,50 +13,64 @@ import { SuperchainConfig } from "src/L1/SuperchainConfig.sol";
+        initialize({
+            _superchainConfig: SuperchainConfig(address(0)),
+            _portal: payable(address(0)),
+            _l2CrossDomainMessenger: Predeploys.L2_CROSS_DOMAIN_MESSENGER
```
should be `address(0)` in the singleton

tynes commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -13,50 +13,64 @@ import { SuperchainConfig } from "src/L1/SuperchainConfig.sol";
+    function initialize(
+        SuperchainConfig _superchainConfig,
+        address payable _portal,
+        address _l2CrossDomainMessenger
```
We always know the address of the l2xdm so we don't need to pass it in here, it only adds an additional footgun. We can hardcode the value to the constant as the arg to `__CrossDomainMessenger_init`

tynes commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -13,50 +13,64 @@ import { SuperchainConfig } from "src/L1/SuperchainConfig.sol";
+
+    /// @notice Address of the OptimismPortal contract.
+    /// @custom:network-specific
+    address payable public portal;
```
Prefer this to be `OptimismPortal` type over `address payable`

tynes commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -13,50 +13,64 @@ import { SuperchainConfig } from "src/L1/SuperchainConfig.sol";
+    /// @return Contract of the OptimismPortal on this chain.
+    /// @custom:legacy
+    function PORTAL() external view returns (OptimismPortal) {
+        return OptimismPortal(portal);
```
this should be able to just return `portal` if we update the type in the storage

tynes commented on packages/contracts-bedrock/src/L1/L1CrossDomainMessenger.sol:
```diff
@@ -13,50 +13,64 @@ import { SuperchainConfig } from "src/L1/SuperchainConfig.sol";
     /// @inheritdoc CrossDomainMessenger
     function _sendMessage(address _to, uint64 _gasLimit, uint256 _value, bytes memory _data) internal override {
-        PORTAL.depositTransaction{ value: _value }(_to, _value, _gasLimit, false, _data);
+        OptimismPortal(portal).depositTransaction{ value: _value }(_to, _value, _gasLimit, false, _data);
```
nit: prefer kwargs style syntax here

tynes commented on packages/contracts-bedrock/src/L2/L2CrossDomainMessenger.sol:
```diff
@@ -15,26 +15,27 @@ import { Constants } from "src/libraries/Constants.sol";
-    constructor(address _l1CrossDomainMessenger) CrossDomainMessenger(_l1CrossDomainMessenger) {
-        initialize();
+    /// @param _l1CrossDomainMessenger Address of the L1CrossDomainMessenger contract on the other network.
+    constructor(address _l1CrossDomainMessenger) CrossDomainMessenger() {
```
no need for a constructor arg here, can just be `address(0)` in the singleton

tynes commented on packages/contracts-bedrock/src/L2/L2CrossDomainMessenger.sol:
```diff
@@ -15,26 +15,27 @@ import { Constants } from "src/libraries/Constants.sol";
-    function initialize() public initializer {
-        __CrossDomainMessenger_init();
+    /// @param _l1CrossDomainMessenger Address of the L1CrossDomainMessenger contract on the other network.
+    function initialize(address _l1CrossDomainMessenger) public initializer {
```
nit: can you update this to be a `CrossDomainMessenger` type instead of `address`?

0xfuturistic replied:
double checking here: this change would clash with the underlying type of `universal/CrossDomainMessenger.sol`. Updating that would have deeper effects. On the other hand, it seems like manually wrapping the address in `CrossDomainMessenger` would require later unwrapping it to use in relation to  `universal/CrossDomainMessenger.sol`, which feels unnecessary. maybe i'm missing something?

tynes commented on packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol:
```diff
@@ -138,10 +135,14 @@ abstract contract CrossDomainMessenger is
 
+    /// @notice Address of the paired CrossDomainMessenger contract on the other chain.
+    /// @custom:network-specific
+    address public otherMessenger;
```
Can you make this be a `CrossDomainMessenger` type? It would be more consistent with other places in the codebase

tynes commented on packages/contracts-bedrock/src/universal/CrossDomainMessenger.sol:
```diff
@@ -352,8 +356,10 @@ abstract contract CrossDomainMessenger is
-    function __CrossDomainMessenger_init() internal onlyInitializing {
+    function __CrossDomainMessenger_init(address _otherMessenger) internal onlyInitializing {
+        otherMessenger = _otherMessenger;
         xDomainMsgSender = Constants.DEFAULT_L2_SENDER;
```
This is touching the same code as https://github.com/ethereum-optimism/optimism/pull/8797 - just to be aware of this, no need to pull in the change here. We may want to consider doing the same check for `otherMessenger` altho I do not think we need to cc @maurelian

tynes commented on packages/contracts-bedrock/test/L1/L1CrossDomainMessenger.t.sol:
```diff
@@ -22,6 +24,46 @@ contract L1CrossDomainMessenger_Test is Bridge_Initializer {
+    }
+
+    /// @dev Tests that the implementation cannot be initialized twice.
+    function test_initialize_cannotInitImpl_reverts() external {
```
This test should be covered in the `Initializable.t.sol`

tynes replied:
Can you be sure to update that file for the modified contracts in this PR?

0xfuturistic replied:
yes absolutely!

0xfuturistic committed:
contracts-bedrock: drop unnecessary parameter in L1CrossDomainMessenger initialize

0xfuturistic committed:
contracts-bedrock: change type of otherMessenger in CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: drop constructor arg for L1CrossDomainMessenger, L2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: fix test for L1CrossDomainMessenger

0xfuturistic committed:
op-bindings: make for CrossdomainMessenger, L1CrossdomainMessenger, L2CrossdomainMessenger

0xfuturistic committed:
op-chain-ops: check fetched addresses from L1CrossDomainMessenger against DeployConfig

0xfuturistic committed:
contracts-bedrock: drop redundant tests for L1CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: drop redundant test for L2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: fix documentation for L2CrossDomainMessenger

0xfuturistic committed:
contracts-bedrock: update slither report

0xfuturistic committed:
contracts-bedrock: update snapshots

0xfuturistic committed:
contracts-bedrock: update semver-lock

0xfuturistic committed:
contracts-bedrock: fix rebase

0xfuturistic committed:
op-bindings: make

tynes commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -107,7 +108,30 @@ func L1CrossDomainMessenger(batch *safe.Batch, implementations superchain.Implem
+			return fmt.Errorf("upgrading L1CrossDomainMessenger: Portal address doesn't match config")
+		}
+
+		if otherMessenger != predeploys.L2CrossDomainMessengerAddr {
```
We can technically not hide this behind `if config != nil`

tynes commented on op-chain-ops/upgrades/l1.go:
```diff
@@ -107,7 +108,30 @@ func L1CrossDomainMessenger(batch *safe.Batch, implementations superchain.Implem
+		}
+	}
+
+	calldata, err := l1CrossDomainMessengerABI.Pack("initialize", superchainConfigProxy, optimismPortal, otherMessenger)
```
This is getting an extra argument, it doesn't need `otherMessenger`

0xfuturistic replied:
great catch. thanks!

tynes commented:
Good after https://github.com/ethereum-optimism/optimism/pull/8631/files#r1448073415 is addressed

0xfuturistic committed:
op-chain-ops: drop unnecessary parameter in L1CrossDomainMessenger L1 upgrade

0xfuturistic committed:
contracts-bedrock: move check out of config existence clause in L1 upgrade script for L1CrossDomainMessenger
